lightning_ir/retrieve/plaid/plaid_indexer.py

```diff
@@ -67,79 +61,94 @@ def add(self, index_batch: IndexBatch, output: BiEncoderOutput) -> None:
             doc_lengths = doc_embeddings.scoring_mask.sum(dim=1)
             embeddings = doc_embeddings.embeddings[doc_embeddings.scoring_mask]
         doc_ids = index_batch.doc_ids
-        embeddings = self.process_embeddings(embeddings)
-
-        if embeddings.shape[0]:
-            if self.residual_codec is None:
-                raise ValueError("Residual codec not trained")
-            codes, residuals = self.residual_codec.compress(embeddings)
-            self.codes.extend(codes.numpy(force=True))
-            self.residuals.extend(residuals.view(-1).numpy(force=True))
 
         self.num_embeddings += embeddings.shape[0]
         self.num_docs += len(doc_ids)
 
         self.doc_lengths.extend(doc_lengths.int().cpu().tolist())
         self.doc_ids.extend(doc_ids)
 
-    def process_embeddings(self, embeddings: torch.Tensor) -> torch.Tensor:
-        """Process embeddings before indexing.
+        if self.index is None:
+            if self._train_embeddings is None:
+                self._train_embeddings = embeddings.new_full(
```

Comment:
We will be removing the train embeddings anyway (see comment above), but aggregating this way creates problems, since any documents with different number of tokens other than the first batch will be thrown out completely. The embeddings are padded with nans and any embeddings with nans are removed later.

Instead, aggregate the embeddings into a list. The example below shows how. Fast plaid converts a tensor into a list of tensors internally anyway, so we are not losing any efficiency.


```suggestion
    # in the __init__ function
    self.embeddings = []

    # in the add function
    self.embeddings.extend(embeddings.split(doc_lengths.int().cpu().tolist()))
```